django/forms/widgets.py

```diff
@@ -151,29 +150,23 @@ def merge(*lists):
         in a certain order. In JavaScript you may not be able to reference a
         global or in CSS you might want to override a style.
         """
-        dependency_graph = defaultdict(set)
-        all_items = OrderedSet()
-        for list_ in filter(None, lists):
-            head = list_[0]
-            # The first items depend on nothing but have to be part of the
-            # dependency graph to be included in the result.
-            dependency_graph.setdefault(head, set())
-            for item in list_:
-                all_items.add(item)
-                # No self dependencies
-                if head != item:
-                    dependency_graph[item].add(head)
+        ts = TopologicalSorter()
+        for head, *tail in filter(None, lists):
+            ts.add(head)  # Ensure that the first items are included.
+            for item in tail:
+                if head != item:  # Avoid circular dependency to self.
+                    ts.add(item, head)
                 head = item
         try:
-            return stable_topological_sort(all_items, dependency_graph)
-        except CyclicDependencyError:
+            return list(ts.static_order())
+        except CycleError:
             warnings.warn(
                 "Detected duplicate Media files in an opposite order: {}".format(
                     ", ".join(repr(list_) for list_ in lists)
                 ),
                 MediaOrderConflictWarning,
             )
-            return list(all_items)
+            return list(dict.fromkeys(chain.from_iterable(filter(None, lists))))
```

Comment:
This is in a fallback branch, so no point generating it in the default non-cycle path...